### PR TITLE
JP-3448: Always flag at least one group when showers are detected

### PR DESCRIPTION
**Pull request by mwregan2**
Resolves [JP-3448](https://jira.stsci.edu/browse/JP-3448)

Closes spacetelescope/jwst#8030

This PR addresses the problem that the shower flagging is not including the flagging of the primary shower as a minimum. The code was updated to match the input parameter which is the time to flag groups after the primary shower. The current default values and code lead to no flagging in MIRI slow mode.

**Checklist**

- [X] added entry in `CHANGES.rst` (either in `Bug Fixes` or `Changes to API`)
- [X] updated relevant tests
- [ ] updated relevant documentation
- [ ] updated relevant milestone(s)
- [x] added relevant label(s)

### CODE REVIEW

File: src/stcal/jump/jump.py [hunk 1]
```diff
@@ -471,8 +473,7 @@ def detect_jumps(
     data /= gain_2d
     err /= gain_2d
     readnoise_2d /= gain_2d
-
-    # Return the updated data quality arrays
+     # Return the updated data quality arrays
```
Comment (kmacdonald-stsci) [inline src/stcal/jump/jump.py:476]: There is an extra space before the comment.

File: src/stcal/jump/jump.py [hunk 2]
```diff
@@ -671,15 +673,20 @@ def extend_ellipses(
...
             sat_pix = np.bitwise_and(gdq_cube[intg, flg_grp, :, :], sat_flag)
             saty, satx = np.where(sat_pix == sat_flag)
             jump_ellipse[saty, satx] = 0
-            gdq_cube[intg, flg_grp, :, :] = np.bitwise_or(gdq_cube[intg, flg_grp, :, :], jump_ellipse)
-    return gdq_cube, num_ellipses
+            out_gdq_cube[intg, flg_grp, :, :] = np.bitwise_or(gdq_cube[intg, flg_grp, :, :], jump_ellipse)
+    diff_cube = out_gdq_cube - gdq_cube
+    return out_gdq_cube, num_ellipses
 
+def find_last_grp(grp, ngrps, num_grps_masked):
```
Comment (kmacdonald-stsci) [inline src/stcal/jump/jump.py:686]: Missing docstring

**Review by kmacdonald-stsci — APPROVED** [review]

File: CHANGES.rst [hunk 3]
```diff
@@ -1,7 +1,10 @@
 1.5.3 (unreleased)
 ==================
 
-- 
+jump
+------------
```
Comment (hbushouse) [inline CHANGES.rst:5]:
```suggestion
----
```

File: CHANGES.rst [hunk 4]
```diff
@@ -1,7 +1,10 @@
...
+Fix the code to at least always flag the group with the shower and the requested
+groups after the primary shower. [#237]
```
Comment (hbushouse) [inline CHANGES.rst:7]:
```suggestion
- Fix the code to at least always flag the group with the shower and the requested
  groups after the primary shower. [#237]
```

**Comment by hbushouse** [comment]
@mwregan2 Still waiting for review comments to be addressed.

**Comment by mwregan2** [comment]
Will do. I need to double check something first.

________________________________

File: src/stcal/jump/jump.py (hunk 1 above)
Comment (mwregan2) [reply src/stcal/jump/jump.py:476]: fixed

File: src/stcal/jump/jump.py (hunk 2 above)
Comment (mwregan2) [reply src/stcal/jump/jump.py]: Added the docstring

File: tests/test_jump.py [hunk 5]
```diff
@@ -352,3 +353,15 @@ def test_calc_num_slices():
...
+def test_find_last_grp():
+    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked=0) == 6)
+    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked=2) == 7)
+    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked=3) == 7)
+    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked=1) == 6)
+    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked=0) == 6)
+    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked=2) == 6)
+    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked=0) == 6)
+    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked=1) == 7)
+    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked=2) == 8)
```
Comment (hbushouse) [inline tests/test_jump.py:367]:
```suggestion
    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked_after=0) == 6)
    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked_after=2) == 7)
    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked_after=3) == 7)
    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked_after=1) == 6)
    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked_after=0) == 6)
    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked_after=2) == 6)
    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked_after=0) == 6)
    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked_after=1) == 7)
    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked_after=2) == 8)
```

File: src/stcal/jump/jump.py [hunk 6]
```diff
@@ -671,15 +673,38 @@ def extend_ellipses(
...
             sat_pix = np.bitwise_and(gdq_cube[intg, flg_grp, :, :], sat_flag)
             saty, satx = np.where(sat_pix == sat_flag)
             jump_ellipse[saty, satx] = 0
-            gdq_cube[intg, flg_grp, :, :] = np.bitwise_or(gdq_cube[intg, flg_grp, :, :], jump_ellipse)
-    return gdq_cube, num_ellipses
+            out_gdq_cube[intg, flg_grp, :, :] = np.bitwise_or(gdq_cube[intg, flg_grp, :, :], jump_ellipse)
+    diff_cube = out_gdq_cube - gdq_cube
+    return out_gdq_cube, num_ellipses
+
+def find_last_grp(grp, ngrps, num_grps_masked_after):
```
Comment (hbushouse) [inline src/stcal/jump/jump.py:686]:
```suggestion
def find_last_grp(grp, ngrps, num_grps_masked):
```

File: src/stcal/jump/jump.py [hunk 7]
```diff
@@ -671,15 +673,38 @@ def extend_ellipses(
...
+    """
+    Parameters
+    ----------
+    grp : int
+        The location of the shower
+    
+    ngrps : int
+        The number of groups in the integration
+    
+    num_grps_masked_after : int
```
Comment (hbushouse) [inline src/stcal/jump/jump.py:696]:
```suggestion
    num_grps_masked : int
```

File: src/stcal/jump/jump.py [hunk 8]
```diff
@@ -671,15 +673,38 @@ def extend_ellipses(
...
+        The requested number of groups to be flagged after the shower
 
+    Returns
+    -------
+    last_grp : int
+        The index of the last group to flag for the shower
+
+    """
+    num_grps_masked_after += 1
+    last_grp = min(grp + num_grps_masked_after, ngrps)
```
Comment (hbushouse) [inline src/stcal/jump/jump.py:706]:
```suggestion
    num_grps_masked += 1
    last_grp = min(grp + num_grps_masked, ngrps)
```

File: tests/test_jump.py [hunk 9]
```diff
@@ -352,3 +353,15 @@ def test_calc_num_slices():
...
+def test_find_last_grp():
+    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked_after=0) == 6)
+    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked_after=2) == 7)
+    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked_after=3) == 7)
+    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked_after=1) == 6)
+    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked_after=0) == 6)
+    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked_after=2) == 6)
+    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked_after=0) == 6)
+    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked_after=1) == 7)
+    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked_after=2) == 8)
```
Comment (hbushouse) [inline tests/test_jump.py:367]:
```suggestion
    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked=0) == 6)
    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked=2) == 7)
    assert (find_last_grp(grp=5, ngrps=7, num_grps_masked=3) == 7)
    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked=1) == 6)
    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked=0) == 6)
    assert (find_last_grp(grp=5, ngrps=6, num_grps_masked=2) == 6)
    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked=0) == 6)
    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked=1) == 7)
    assert (find_last_grp(grp=5, ngrps=8, num_grps_masked=2) == 8)
```

**Comment by hbushouse** [comment]
I tried fixing obvious errors (like mismatching argument names) to get this PR to pass CI tests, but it's still failing CI's with both differences in unit test results and a real error "UnboundLocalError: cannot access local variable 'total_showers' where it is not associated with a value". So something still needs fixing in the basic code.

**Review by hbushouse — APPROVED** [review]
Looks good now and all CI tests are passing.

**Pull request merged**